src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs

```diff
@@ -1264,28 +1283,139 @@ public static bool IsPatternContext(this SyntaxTree syntaxTree, SyntaxToken left
             {
                 return true;
             }
 
             // e is ($$
             // e is (..., $$
-            if (leftToken.IsKind(SyntaxKind.OpenParenToken, SyntaxKind.CommaToken) && leftToken.Parent.IsKind(SyntaxKind.PositionalPatternClause))
+            if (leftToken.IsKind(SyntaxKind.OpenParenToken, SyntaxKind.CommaToken) && node.IsKind(SyntaxKind.PositionalPatternClause))
             {
                 return true;
             }
 
             // e is { P: $$
             // e is { ..., P: $$
-            if (leftToken.IsKind(SyntaxKind.ColonToken) && leftToken.Parent.IsKind(SyntaxKind.NameColon) &&
-                leftToken.Parent.IsParentKind(SyntaxKind.Subpattern))
+            if (leftToken.IsKind(SyntaxKind.ColonToken) && node.IsKind(SyntaxKind.NameColon) &&
+                node.IsParentKind(SyntaxKind.Subpattern))
+            {
+                return true;
+            }
+
+            // e is { P: ($$
+            if (leftToken.IsKind(SyntaxKind.OpenParenToken) && node.IsParentKind(SyntaxKind.ConstantPattern))
+            {
+                return true;
+            }
+
+#if !CODE_STYLE
+            // if (e is { P: (1 or $$
+            if ((leftToken.IsKind(SyntaxKind.AndKeyword) || leftToken.IsKind(SyntaxKind.OrKeyword)) &&
+                node is BinaryPatternSyntax)
             {
                 return true;
             }
 
+            if (node.IsParentKind(SyntaxKind.IsPatternExpression))
+            {
+                // e is 1 and $$
+                if (node.IsKind(SyntaxKind.AndPattern))
+                {
+                    return true;
+                }
+
+                // e is 1 or $$
+                if (node.IsKind(SyntaxKind.OrPattern))
+                {
+                    return true;
+                }
+
+                // e is not $$
+                if (node.IsKind(SyntaxKind.NotPattern))
+                {
+                    return true;
+                }
+
+                // e is ($$ 1 or 2)
+                if (node.IsKind(SyntaxKind.ParenthesizedPattern))
+                {
+                    return true;
+                }
+            }
+#endif
+
+            return false;
+        }
+
+        public static bool IsAtEndOfPattern(this SyntaxTree syntaxTree, SyntaxToken leftToken, int position)
+        {
+#if !CODE_STYLE
+            leftToken = leftToken.GetPreviousTokenIfTouchingWord(position);
+
+            // This check should cover the majority of cases, e.g.:
+            // e is 1 $$
+            // e is >= 0 $$
+            // e is { P: (1 $$
+            if (leftToken.IsLastTokenOfNode<PatternSyntax>(out var patternSyntax, includeZeroWidth: false))
+            {
+                // Patterns such as 'e is not $$', 'e is 1 or $$', and 'e is ($$' should be invalid here.
+                if (!(patternSyntax is UnaryPatternSyntax || patternSyntax is BinaryPatternSyntax || patternSyntax is RecursivePatternSyntax))
```

Comment:
i don't understand this.  it doesn't seem to be true at all.  For example:

```c#
e is 1 or 2 $$
```

We'll have a BinaryPattern, and we'll be at the end of it.  And that makes sense since you could write:  `1 or 2 or 3`

Reply:
I do agree that the example you provided is valid. The comment was referring to jumbled statements such as 'e is not and', 'e is 1 or and', etc. being invalid.